### PR TITLE
Fixes the Issue #1164: i.e. Download of files with duplicate names

### PR DESCRIPTION
**Pull request by rajul**
Fixes the Issue #1164

### CODE REVIEW

File: sunpy/net/download.py [hunk 1]
```diff
@@ -58,6 +58,7 @@ def _start_download(self, url, path, callback, errback):
                         else:
                             fd.write(rec)
         except Exception, e:
+            # TODO: Fix the silent failing
```
Comment (dpshelio) [inline sunpy/net/download.py]: Do we have an issue already about this? Should this be here?

File: sunpy/database/database.py [hunk 2]
```diff
@@ -301,27 +307,34 @@ def _download_and_collect_entries(self, query_result, client=None,
...
+            else:
+                raise ValueError('The path is neither a file nor directory')
+
+            for entry in entries:
+                entry.source = qr_entry.source
+                entry.provider = qr_entry.provider
+                entry.physobs = qr_entry.physobs
+                entry.fileid = qr_entry.fileid
+                entry.observation_time_start = \
+                qr_entry.observation_time_start
```
Comment (dpshelio) [inline sunpy/database/database.py]: would this be the right indentation?

File: sunpy/database/tests/test_database.py [hunk 3]
```diff
@@ -785,6 +787,40 @@ def test_fetch(database, download_query, tmpdir):
...
+
+    db.fetch(*download_query, path=path)
+    
+    # Test
+    assert len(db) == 2
+    
+    dir_contents = os.listdir(tmp_test_dir)
+    assert dir_contents[-2] == 'aia_lev1_335a_2012_08_05t00_00_02_62z_image_lev1_fits.fits'
+    assert dir_contents[-1] == 'aia_lev1_94a_2012_08_05t00_00_01_12z_image_lev1_fits.fits'
+    assert dir_contents[-1].split('.')[0] != dir_contents[-2].split('.')[1]
```
Comment (dpshelio) [inline sunpy/database/tests/test_database.py]: Though in my machine the test runs successfully, it doesn't in Travis... it seems that the order of the `dir_contents` it doesn't match... and [https://travis-ci.org/sunpy/sunpy/jobs/54440291#L1278](therefore compares 94 with 335). Could it be that `os.listdir` does not provide always the same order?  I've just made a test in my machine and I get this:
```
 'efz20000913_063611',
 'efz20000912_212410',
 'efz20000913_170011',
```
which it doesn't follow any logical order.  Would it better to use some regex? and see whether the file is there? or.. just check whether it exists:
```python
assert os.path.is_file(os.path.join(tmp_test_dir, 'aia_lev1_335a_2012_08_05t00_00_02_62z_image_lev1_fits.fits'))
```

File: sunpy/net/download.py (hunk 1 above)
Comment (rajul) [inline sunpy/net/download.py]: Yes we have an Issue about this one..... #1331

File: sunpy/database/database.py (hunk 2 above)
Comment (rajul) [inline sunpy/database/database.py]: Fixed this....thanks!!

File: sunpy/database/tests/test_database.py (hunk 3 above)
Comment (rajul) [inline sunpy/database/tests/test_database.py]: I looked at the errors in travis.....the test I added is passing on travis too.....but the ones which are failing are the heuristic tests in test_mapbase.py.......I will open an issue for that and then fix them as ones in test_callisto.py ( #1332)....using asset_allclose....

Comment (dpshelio) [inline sunpy/database/tests/test_database.py]: Nooo... it's not passing.  Look at the 'online' run, this test is not passing.

Comment (rajul) [inline sunpy/database/tests/test_database.py]: @dpshelio oh yes...you are right....i was not looking at the 'online' build.....great will fix this as you have suggested above....thanks!!

**Pull request closed without merging**